// react-storefront-amp/search/AmpSearchField.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = AmpSearchField;
exports.styles = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _SearchField = _interopRequireDefault(require("react-storefront/search/SearchField"));

var _DataBindingContext = _interopRequireDefault(require("../bind/DataBindingContext"));

var _AmpSearchSubmitButton = _interopRequireDefault(require("./AmpSearchSubmitButton"));

var _makeStyles = _interopRequireDefault(require("@material-ui/core/styles/makeStyles"));

var _amp = require("next/amp");

var styles = function styles(theme) {
  return {
    hidden: {
      display: 'block'
    }
  };
};

exports.styles = styles;
var useStyles = (0, _makeStyles["default"])(styles, {
  name: 'RSFSearchField'
});

function AmpSearchField(props) {
  var amp = (0, _amp.useAmp)();

  var _useContext = (0, _react.useContext)(_DataBindingContext["default"]),
      ampState = _useContext.ampState;

  var classes = useStyles(props);
  return _react["default"].createElement("div", null, _react["default"].createElement(_SearchField["default"], (0, _extends2["default"])({}, props, {
    classes: amp ? classes : undefined,
    SubmitButtonComponent: _AmpSearchSubmitButton["default"],
    inputProps: amp ? {
      'amp-bind': "value->".concat(ampState, ".text || ''"),
      on: "input-debounced:AMP.setState({ ".concat(ampState, ": { text: ").concat(ampState, ".___moov_submitting ? ").concat(ampState, ".text : event.value } })")
    } : undefined,
    clearButtonProps: amp ? {
      on: "tap:AMP.setState({ ".concat(ampState, ": { text: '' }})"),
      'amp-bind': "hidden->".concat(ampState, ".text.length == 0")
    } : undefined,
    submitButtonProps: amp ? {
      'amp-bind': "hidden->".concat(ampState, ".text.length == 0")
    } : undefined
  })));
}

AmpSearchField.propTypes = _SearchField["default"].propTypes;
AmpSearchField.defaultProps = _SearchField["default"].defaultProps;
