// react-storefront-amp/search/AmpSearchDrawer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = AmpSearchDrawer;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _SearchDrawer = _interopRequireDefault(require("react-storefront/search/SearchDrawer"));

var _DataBindingProvider = _interopRequireDefault(require("../bind/DataBindingProvider"));

var _AmpDrawer = _interopRequireDefault(require("../AmpDrawer"));

var _amp = require("next/amp");

var _head = _interopRequireDefault(require("next/head"));

function AmpSearchDrawer(props) {
  return _react["default"].createElement(_react["default"].Fragment, null, (0, _amp.useAmp)() && _react["default"].createElement(_head["default"], null, _react["default"].createElement("script", {
    async: true,
    "custom-element": "amp-form",
    src: "https://cdn.ampproject.org/v0/amp-form-0.1.js"
  })), _react["default"].createElement(_DataBindingProvider["default"], {
    id: "rsfSearchDrawer",
    store: {
      open: false,
      text: ''
    },
    root: null,
    updateStore: Function.prototype
  }, _react["default"].createElement(_SearchDrawer["default"], (0, _extends2["default"])({}, props, {
    DrawerComponent: _AmpDrawer["default"]
  }))));
}

AmpSearchDrawer.propTypes = {};
AmpSearchDrawer.defaultProps = {};
