// react-storefront-amp/search/AmpSearchSuggestions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = AmpSearchSuggestions;
exports.styles = void 0;

var _react = _interopRequireWildcard(require("react"));

var _makeStyles = _interopRequireDefault(require("@material-ui/core/styles/makeStyles"));

var _SearchSuggestions = _interopRequireDefault(require("react-storefront/search/SearchSuggestions"));

var _SearchSuggestionItem = _interopRequireDefault(require("react-storefront/search/SearchSuggestionItem"));

var _SearchSuggestionGroup = _interopRequireDefault(require("react-storefront/search/SearchSuggestionGroup"));

var _head = _interopRequireDefault(require("next/head"));

var _DataBindingContext = _interopRequireDefault(require("../bind/DataBindingContext"));

var _amp = require("next/amp");

var styles = function styles(theme) {
  return {
    root: {},
    thumbnailGroup: {},
    group: {},
    container: {
      position: 'relative',
      height: '100%',
      overflowY: 'auto'
    }
  };
};

exports.styles = styles;
var useStyles = (0, _makeStyles["default"])(styles, {
  name: 'RSFAmpSearchSuggestions'
});

function AmpSearchSuggestions(props) {
  var classes = useStyles(props);

  var _useContext = (0, _react.useContext)(_DataBindingContext["default"]),
      ampState = _useContext.ampState;

  if (!(0, _amp.useAmp)()) return _react["default"].createElement(_SearchSuggestions["default"], props);
  return _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement(_head["default"], null, _react["default"].createElement("script", {
    async: true,
    "custom-element": "amp-list",
    src: "https://cdn.ampproject.org/v0/amp-list-0.1.js"
  }), _react["default"].createElement("script", {
    async: true,
    "custom-template": "amp-mustache",
    src: "https://cdn.ampproject.org/v0/amp-mustache-0.2.js"
  })), _react["default"].createElement("div", {
    className: classes.container
  }, _react["default"].createElement("amp-list", {
    layout: "fill",
    "class": classes.root,
    "amp-bind": "src->\"/api/suggestions?q=\" + (".concat(ampState, ".text ? encodeURIComponent(").concat(ampState, ".text) : '')"),
    items: "groups",
    "reset-on-refresh": true
  }, _react["default"].createElement("template", {
    type: "amp-mustache"
  }, _react["default"].createElement(_SearchSuggestionGroup["default"], {
    caption: "{{caption}}",
    ui: "{{ui}}",
    "amp-bind": "class->{{thumbnails}} ? \"".concat(classes.thumbnailGroup, "\" : \"").concat(classes.group, "\"")
  }, '{{#links}}', _react["default"].createElement(_SearchSuggestionItem["default"], {
    thumbnailProps: {
      height: 100,
      width: 100
    },
    ui: "{{ui}}",
    item: {
      as: '{{as}}',
      href: '{{href}}',
      text: '{{text}}',
      thumbnail: {
        src: '{{thumbnail.src}}',
        alt: '{{thumbnail.alt}}'
      }
    }
  }), '{{/links}}')))));
}

AmpSearchSuggestions.propTypes = _SearchSuggestions["default"].propTypes;
AmpSearchSuggestions.defaultProps = _SearchSuggestions["default"].defaultProps;
