// react-storefront-amp/carousel/AmpCarouselThumbnails.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _clsx4 = _interopRequireDefault(require("clsx"));

var _AmpImage = _interopRequireDefault(require("../AmpImage"));

var _CarouselThumbnails = _interopRequireWildcard(require("react-storefront/carousel/CarouselThumbnails"));

var _styles = require("@material-ui/core/styles");

var _amp = require("next/amp");

var _withDataBinding = _interopRequireDefault(require("../bind/withDataBinding"));

/**
 * @license
 * Copyright © 2017-2018 Moov Corporation.  All rights reserved.
 */
var useStyles = (0, _styles.makeStyles)(_CarouselThumbnails.styles, {
  name: 'RSFAmpCarouselThumbnails'
});

function AmpCarouselThumbnails(props) {
  var _clsx;

  if (!(0, _amp.useAmp)()) return _react["default"].createElement(_CarouselThumbnails["default"], props);
  var amp = props.amp,
      carouselId = props.carouselId,
      thumbnails = props.thumbnails,
      className = props.className,
      thumbnailPosition = props.thumbnailPosition;
  var styles = useStyles();
  var isVertical = ['left', 'right'].includes(thumbnailPosition);
  return _react["default"].createElement("div", {
    className: (0, _clsx4["default"])(className, styles.thumbs, styles.tabsRoot, (_clsx = {}, (0, _defineProperty2["default"])(_clsx, styles.tabsVertical, isVertical), (0, _defineProperty2["default"])(_clsx, styles.tabsRootLeft, thumbnailPosition === 'left'), (0, _defineProperty2["default"])(_clsx, styles.tabsRootRight, thumbnailPosition === 'right'), (0, _defineProperty2["default"])(_clsx, styles.tabsRootTop, thumbnailPosition === 'top'), (0, _defineProperty2["default"])(_clsx, styles.tabsRootBottom, thumbnailPosition === 'bottom'), _clsx))
  }, thumbnails.map(function (props, i) {
    var icon = _react["default"].createElement(_AmpImage["default"], (0, _extends2["default"])({}, props, {
      ImgElement: "amp-img",
      layout: "fill",
      contain: true,
      className: styles.thumb
    }));

    return _react["default"].createElement("div", (0, _extends2["default"])({
      className: (0, _clsx4["default"])(styles.tabRoot, i === 0 && styles.selectedTab),
      on: "tap:".concat(carouselId, ".goToSlide(index=").concat(i, ")")
    }, amp.bind({
      attribute: 'class',
      value: "".concat(amp.getValue(), " == ").concat(i, " ? '").concat((0, _clsx4["default"])(styles.tabRoot, styles.selectedTab), "' : '").concat(styles.tabRoot, "'")
    })), _react["default"].createElement("div", {
      className: (0, _clsx4["default"])(styles.tabWrapper)
    }, icon), _react["default"].createElement("div", (0, _extends2["default"])({
      className: (0, _clsx4["default"])(styles.tabsIndicator, (0, _defineProperty2["default"])({}, styles.hidden, i !== 0))
    }, amp.bind({
      attribute: 'class',
      value: "".concat(amp.getValue(), " == ").concat(i, " ? '").concat((0, _clsx4["default"])(styles.tabsIndicator), "' : '").concat((0, _clsx4["default"])(styles.tabsIndicator, (0, _defineProperty2["default"])({}, styles.hidden, true)), "'")
    }))));
  }));
}

var _default = (0, _withDataBinding["default"])(AmpCarouselThumbnails);

exports["default"] = _default;
