// react-storefront-amp/carousel/AmpMediaCarousel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _DataBindingProvider = _interopRequireDefault(require("../bind/DataBindingProvider"));

var _AmpCarousel = _interopRequireDefault(require("./AmpCarousel"));

var _AmpCarouselThumbnails = _interopRequireDefault(require("./AmpCarouselThumbnails"));

var _AmpDrawer = _interopRequireDefault(require("../AmpDrawer"));

var _AmpMedia = _interopRequireDefault(require("./AmpMedia"));

var _MediaCarousel = _interopRequireDefault(require("react-storefront/carousel/MediaCarousel"));

var _styles = require("@material-ui/core/styles");

var _amp = require("next/amp");

var _withDataBinding = _interopRequireDefault(require("../bind/withDataBinding"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var styles = function styles(theme) {
  return {
    lightbox: {
      height: '100%',
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      '& > div:first-of-type': (0, _defineProperty2["default"])({}, theme.breakpoints.down('xs'), {
        margin: 0
      })
    },
    thumbnails: {
      paddingBottom: theme.spacing(2),
      order: 'inherit'
    },
    lightboxMagnifyHint: {
      display: 'none'
    }
  };
};

var useStyles = (0, _styles.makeStyles)(styles, {
  name: 'RSFAmpMediaCarousel'
});

function AmpMediaCarousel(_ref) {
  var amp = _ref.amp,
      _ref$maxItems = _ref.maxItems,
      maxItems = _ref$maxItems === void 0 ? 8 : _ref$maxItems,
      props = (0, _objectWithoutProperties2["default"])(_ref, ["amp", "maxItems"]);
  if (!(0, _amp.useAmp)()) return _react["default"].createElement(_MediaCarousel["default"], props);
  var classes = useStyles({
    classes: props.classes
  });

  function createSpace(type, magnify) {
    return function (_, index) {
      if (index >= props.media.full.length) return null;
      var magnifyPrefix = "".concat(amp.getValue('media'), ".full[").concat(index, "].magnify.src || ");
      return _objectSpread({}, props.media[type][index], {}, amp.bind({
        attribute: 'src',
        value: "".concat(magnify ? magnifyPrefix : '').concat(amp.getValue('media'), ".").concat(type, "[").concat(index, "].src")
      }));
    };
  }

  function createMedia(isMagnify) {
    return {
      full: Array(maxItems).fill(0).map(createSpace('full', isMagnify)).filter(Boolean),
      thumbnails: Array(maxItems).fill(0).map(createSpace('thumbnails')).filter(Boolean)
    };
  }

  var mediaSpace = createMedia(false);
  var lightboxMediaSpace = createMedia(true);
  return _react["default"].createElement(_DataBindingProvider["default"], {
    id: "rsfMediaCarousel",
    store: {
      open: false
    },
    root: null,
    updateStore: Function.prototype
  }, _react["default"].createElement(_MediaCarousel["default"], (0, _extends2["default"])({}, props, {
    id: "ampCarousel",
    pairedCarouselId: "ampLightboxCarousel",
    media: mediaSpace,
    CarouselComponent: _AmpCarousel["default"],
    MediaComponent: _AmpMedia["default"],
    CarouselThumbnailsComponent: _AmpCarouselThumbnails["default"],
    bind: {
      index: "ampCarousel.index",
      pairedIndex: "ampLightboxCarousel.index",
      open: 'open'
    }
  })), _react["default"].createElement(_AmpDrawer["default"], {
    open: true,
    showCloseButton: true,
    fullscreen: true
  }, _react["default"].createElement("div", {
    className: classes.lightbox
  }, _react["default"].createElement(_MediaCarousel["default"], (0, _extends2["default"])({}, props, {
    id: "ampLightboxCarousel",
    pairedCarouselId: "ampCarousel",
    media: lightboxMediaSpace,
    CarouselComponent: _AmpCarousel["default"],
    MediaComponent: _AmpMedia["default"],
    CarouselThumbnailsComponent: _AmpCarouselThumbnails["default"],
    thumbsClassName: classes.thumbnails,
    className: props.lightboxClassName,
    magnifyHintClassName: classes.lightboxMagnifyHint,
    thumbnailPosition: props.thumbnailPosition === 'top' ? 'top' : 'bottom',
    bind: {
      index: "ampLightboxCarousel.index",
      pairedIndex: "ampCarousel.index"
    }
  })))));
}

AmpMediaCarousel.propTypes = _MediaCarousel["default"].propTypes;
AmpMediaCarousel.defaultProps = _objectSpread({}, _MediaCarousel["default"].defaultProps, {
  id: 'carousel'
});

var _default = (0, _withDataBinding["default"])(AmpMediaCarousel);

exports["default"] = _default;
