// react-storefront-analytics/AnalyticsProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _analytics = require("./analytics");

var _AnalyticsContext = _interopRequireDefault(require("./AnalyticsContext"));

var _amp = require("next/amp");

/**
 * @license
 * Copyright © 2017-2019 Moov Corporation.  All rights reserved.
 */

/**
 * Components that are decendents of AnalyticsProvider can use the `useAnalytics` hook to:
 * 1) Get access to the `fire` function which can be used to broadcast analytics events:
 *
 * ```
 *  import React, { useCallback } from 'react'
 *  import Button from '@material-ui/core/Button'
 *  import { useAnalytics } from 'react-storefront-analytics'
 *
 *  const MyComponent = () => {
 *    const { fire } = useAnalytics()
 *
 *    // This will call the someEvent() method on all configured analytics targets.
 *    const fireAnalyticsEvent = useCallback(() => {
 *      const eventData = { foo: 'bar' }
 *      fire('someEvent', eventData)
 *    }, [])
 *
 *    return <Button onClick={fireAnalyticsEvent}>Click Me</Button>
 *  }
 * ```
 *
 * 2) Get access to the `events` broadcaster object to listen to any event being triggered:
 * ```
 *  import { useAnalytics } from 'react-storefront-analytics'
 *
 *  const MyAnalyticsHandler = () => {
 *    const { events } = useAnalytics()
 *    events.on('someEvent', e => console.log('someEvent triggered with: ', e.eventParams))
 *  }
 * ```
 * `useAnalytics.events` can be used to implement analytics tracking for any vendor.
 *
 * 3) Get access to `addAmpAnalyticsTag` function. This function can be used to build the
 * <amp-analytics> tag in a custom analytics handler. `addAmpAnalyticsTag` takes a function
 * as a parameter. The function provided is called with an array of all targets that are
 * tracked by being wrapped by a <Track> component. Each target is represented as an object
 * in the form of `{ event, eventParams, selector }` where `event` represents the name of the
 * event, `eventParams` represents the event data passed as props to <Track> and the selector
 * is a unique auto-generated CSS selector pointing to a specific target.
 * ```
 * import { useAmp } from 'next/amp'
 * import { useAnalytics } from 'react-storefront-analytics'
 *
 * const MyAnalyticsHandler = () => {
 *   const { events, addAmpAnalyticsTag } = useAnalytics()
 *
 *   if (useAmp()) {
 *     let triggers = []
 *     addAmpAnalyticsTag(targets => {
 *       targets.map(({event, eventParams, selector}) => {
 *         triggers.push({on: 'click', request: 'event', vars: {id: eventParams.id}, selector})
 *       })
 *       const configuration = {triggers, vars: {myAmpVar: 'myAmpVar'}}
 *       return (
 *         `<amp-analytics type="myVendorType">` +
 *         `<script type="application/json">${JSON.stringify(configuration)}</script>` +
 *         `</amp-analytics>`
 *       )
 *     })
 *     return null
 *   } else {
 *     *handle analytics events for non-AMP*
 *     events.on(...)
 *   }
 * }
 * ```
 */
var AnalyticsProvider = function AnalyticsProvider(_ref) {
  var children = _ref.children,
      delayUntilInteractive = _ref.delayUntilInteractive;

  // Analytics state and AMP handlers are stateful in module scope
  if ((0, _amp.useAmp)()) {
    (0, _analytics.reset)();
  }

  (0, _react.useEffect)(function () {
    // Allow targets to add event listeners before initializing
    setTimeout(function () {
      (0, _analytics.initialize)(delayUntilInteractive);
    });
  }, []);
  var value = (0, _react.useMemo)(function () {
    return {
      fire: _analytics.fire,
      events: _analytics.events,
      addAmpAnalyticsTag: _analytics.addAmpAnalyticsTag
    };
  }, []);
  return _react["default"].createElement(_AnalyticsContext["default"].Provider, {
    value: value
  }, children);
};

AnalyticsProvider.propTypes = {
  /**
   * Set to true to delay loading of analytics until the app is interactive
   */
  delayUntilInteractive: _propTypes["default"].bool
};
AnalyticsProvider.defaultProps = {
  trackPageViews: true
};
var _default = AnalyticsProvider;
exports["default"] = _default;
