// react-storefront-analytics/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AnalyticsProvider", {
  enumerable: true,
  get: function get() {
    return _AnalyticsProvider["default"];
  }
});
Object.defineProperty(exports, "Track", {
  enumerable: true,
  get: function get() {
    return _Track["default"];
  }
});
Object.defineProperty(exports, "useAnalytics", {
  enumerable: true,
  get: function get() {
    return _useAnalytics["default"];
  }
});
Object.defineProperty(exports, "GoogleAnalytics", {
  enumerable: true,
  get: function get() {
    return _GoogleAnalytics["default"];
  }
});
Object.defineProperty(exports, "GoogleTagManager", {
  enumerable: true,
  get: function get() {
    return _GoogleTagManager["default"];
  }
});
Object.defineProperty(exports, "FirebasePerformanceMonitoring", {
  enumerable: true,
  get: function get() {
    return _FirebasePerformanceMonitoring["default"];
  }
});
Object.defineProperty(exports, "TrackPageView", {
  enumerable: true,
  get: function get() {
    return _TrackPageView["default"];
  }
});
Object.defineProperty(exports, "renderAmpAnalyticsTags", {
  enumerable: true,
  get: function get() {
    return _analytics.renderAmpAnalyticsTags;
  }
});

var _AnalyticsProvider = _interopRequireDefault(require("./AnalyticsProvider.js"));

var _Track = _interopRequireDefault(require("./Track.js"));

var _useAnalytics = _interopRequireDefault(require("./useAnalytics.js"));

var _GoogleAnalytics = _interopRequireDefault(require("./targets/GoogleAnalytics.js"));

var _GoogleTagManager = _interopRequireDefault(require("./targets/GoogleTagManager.js"));

var _FirebasePerformanceMonitoring = _interopRequireDefault(require("./firebase/FirebasePerformanceMonitoring.js"));

var _TrackPageView = _interopRequireDefault(require("./TrackPageView.js"));

var _analytics = require("./analytics.js");
