// react-storefront-amp/menu/AmpMenuDrawer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = AmpMenuDrawer;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _head = _interopRequireDefault(require("next/head"));

var _AmpMenuBody = _interopRequireDefault(require("./AmpMenuBody"));

var _styles = require("@material-ui/core/styles");

function AmpMenuDrawer(_ref) {
  var id = _ref.id,
      classes = _ref.classes,
      drawerWidth = _ref.drawerWidth,
      anchor = _ref.anchor,
      cards = _ref.cards,
      rootHeader = _ref.rootHeader,
      rootFooter = _ref.rootFooter;

  var styles = function styles(theme) {
    return {
      root: {
        position: 'relative',
        maxWidth: '100vw',
        marginTop: theme.headerHeight,
        height: "calc(100vh - ".concat(theme.headerHeight, "px)"),
        backgroundColor: theme.palette.background.paper,
        boxShadow: '10px 2px 10px -5px rgba(0, 0, 0, 0.2)',
        paddingBottom: '64px',
        width: drawerWidth,
        '& [expanded] > h3': {
          color: theme.palette.secondary.contrastText,
          backgroundColor: theme.palette.secondary.main,
          borderColor: theme.palette.secondary.main,
          '& svg': {
            fill: theme.palette.secondary.contrastText
          }
        },
        '& a': {
          color: theme.typography.body1.color
        },
        '& ~ div[class*="amphtml-sidebar-mask"]': {
          background: 'rgba(0, 0, 0, 0.5)',
          opacity: 1
        }
      }
    };
  };

  classes = (0, _styles.makeStyles)(styles, {
    name: 'RSFAmpMenuDrawer'
  })({
    classes: classes
  });
  return _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement(_head["default"], null, _react["default"].createElement("script", {
    async: true,
    "custom-element": "amp-sidebar",
    src: "https://cdn.ampproject.org/v0/amp-sidebar-0.1.js"
  }), _react["default"].createElement("script", {
    async: true,
    "custom-element": "amp-bind",
    src: "https://cdn.ampproject.org/v0/amp-bind-0.1.js"
  })), _react["default"].createElement("amp-sidebar", {
    id: id,
    key: "sidebar",
    "class": classes.root,
    layout: "nodisplay",
    side: anchor,
    on: "sidebarClose:AMP.setState({ rsfMenuState: { open: false } })"
  }, _react["default"].createElement(_AmpMenuBody["default"], {
    cards: cards,
    rootHeader: rootHeader,
    rootFooter: rootFooter,
    drawerWidth: drawerWidth
  })));
}

AmpMenuDrawer.propTypes = {
  id: _propTypes["default"].string.isRequired,
  anchor: _propTypes["default"].string
};
AmpMenuDrawer.defaultProps = {
  id: 'rsfMenu',
  anchor: 'left'
};
