// react-storefront-amp/menu/AmpMenuBranch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = AmpMenuBranch;
exports.styles = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _MenuBranch = _interopRequireDefault(require("react-storefront/menu/MenuBranch"));

var _makeStyles = _interopRequireDefault(require("@material-ui/core/styles/makeStyles"));

var styles = function styles(theme) {
  return {
    root: {
      '& section': {
        borderBottom: 'none'
      }
    }
  };
};

exports.styles = styles;
var useStyles = (0, _makeStyles["default"])(styles, {
  name: 'RSFAmpMenuBranch'
});
/**
 * Overrides MenuBranch, adding a tap handler that updates the AMP state
 * to show the appropriate card.
 */

function AmpMenuBranch(props) {
  var item = props.item;

  var _useStyles = useStyles(props),
      root = _useStyles.root,
      classes = (0, _objectWithoutProperties2["default"])(_useStyles, ["root"]);

  var onTapState = JSON.stringify({
    rsfMenuState: (0, _defineProperty2["default"])({
      card: item._rsfMenuCard
    }, "card".concat(item._rsfMenuCard), item._rsfMenuItem)
  });
  return _react["default"].createElement("div", {
    className: root
  }, _react["default"].createElement(_MenuBranch["default"], (0, _extends2["default"])({}, props, {
    classes: classes,
    itemProps: {
      on: "tap:AMP.setState(".concat(onTapState, ")")
    }
  })));
}

AmpMenuBranch.propTypes = _MenuBranch["default"].propTypes;
AmpMenuBranch.defaultProps = _MenuBranch["default"].defaultProps;
