// react-storefront-amp/menu/AmpMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = AmpMenu;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _Menu = _interopRequireDefault(require("react-storefront/menu/Menu"));

var _AmpMenuDrawer = _interopRequireDefault(require("./AmpMenuDrawer"));

var _amp = require("next/amp");

var _AmpState = _interopRequireDefault(require("../AmpState"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 *
 * Overrides Menu and renders all menu items at once, using CSS to show/hide then based
 * on the user's navigation through the menu.
 */
function AmpMenu(props) {
  if (!(0, _amp.useAmp)()) {
    return _react["default"].createElement(_Menu["default"], props);
  }

  var cards = [];
  var root = props.root;
  var nextId = 0;

  function addMenuData(item) {
    var depth = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;

    var result = _objectSpread({}, item, {
      _rsfMenuItem: nextId++,
      _rsfMenuCard: depth,
      root: depth === 0,
      items: item && item.items && item.items.map(function (item) {
        return addMenuData(item, depth + 1);
      })
    });

    var card = cards[depth];
    if (!card) card = cards[depth] = [];
    card.push(result);
    return result;
  }

  root = addMenuData(props.root);
  return _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement(_AmpState["default"], {
    id: "rsfMenuState",
    initialState: {
      card0: 0,
      card: 0
    }
  }), _react["default"].createElement(_Menu["default"], (0, _extends2["default"])({}, props, {
    renderDrawer: function renderDrawer() {
      return _react["default"].createElement(_AmpMenuDrawer["default"], (0, _extends2["default"])({}, props, {
        root: root,
        cards: cards
      }));
    }
  })));
}

AmpMenu.propTypes = _Menu["default"].propTypes;
AmpMenu.defaultProps = _Menu["default"].defaultProps;
