// react-storefront-amp/menu/AmpMenuCard.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = AmpMenuCard;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _MenuCard = _interopRequireDefault(require("react-storefront/menu/MenuCard"));

var _MenuContext = _interopRequireDefault(require("react-storefront/menu/MenuContext"));

var _amp = require("next/amp");

function AmpMenuCard(props) {
  if (!(0, _amp.useAmp)()) return _react["default"].createElement(_MenuCard["default"], props);
  var item = props.item;

  var _useContext = (0, _react.useContext)(_MenuContext["default"]),
      drawerWidth = _useContext.drawerWidth;

  return _react["default"].createElement("div", {
    style: {
      width: drawerWidth
    }
  }, item.map(function (item, i) {
    return _react["default"].createElement("div", {
      key: i,
      "amp-bind": "hidden->rsfMenuState.card".concat(item._rsfMenuCard, " != ").concat(item._rsfMenuItem)
    }, _react["default"].createElement(_MenuCard["default"], (0, _extends2["default"])({}, props, {
      item: item,
      headerProps: {
        backButtonProps: {
          on: 'tap:AMP.setState({ rsfMenuState: { card: rsfMenuState.card - 1 }})'
        }
      }
    })));
  }));
}
