// react-storefront-analytics/firebase/FirebasePerformanceMonitoring.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = FirebasePerformanceMonitoring;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reportPerformanceMetrics = _interopRequireDefault(require("./reportPerformanceMetrics"));

var _router = require("next/router");

var _FirebaseContext = _interopRequireDefault(require("./FirebaseContext"));

var _amp = require("next/amp");

/* istanbul ignore else */
if (typeof window !== 'undefined') {
  window.__rsfFirebaseSendPerformanceMetrics = _reportPerformanceMetrics["default"];
}
/**
 * Reports RUM metrics to Firebase Performance Monitoring.
 *
 * Your firebase config object must be provided as the `config` prop. To get your firebase config object:
 *
 * 1.) Go to your the Settings icon Project settings in the Firebase console.
 * 2.) In the Your apps card, select the nickname of the app for which you need a config object.
 * 3.) Select Config from the Firebase SDK snippet pane.
 * 4.) Provide the config object using the `config` prop on this component.
 */


function FirebasePerformanceMonitoring(_ref) {
  var config = _ref.config,
      firebaseSdkUrl = _ref.firebaseSdkUrl,
      clientSideNavigationTraceName = _ref.clientSideNavigationTraceName,
      children = _ref.children;
  var context = (0, _react.useMemo)(function () {
    return {
      trace: null
    };
  }, []);
  (0, _react.useEffect)(function () {
    if (config) {
      _router.Router.events.on('routeChangeStart', function () {
        context.trace = window.firebasePerf.trace(clientSideNavigationTraceName);
        context.trace.start(); // the trace will be ended by FirebaseNavigationTrace
      });
    }
  }, [config]);
  var scripts = null;

  if (config && !(0, _amp.useAmp)()) {
    scripts = _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement("link", {
      href: "https://firebaselogging.googleapis.com",
      rel: "preconnect",
      crossOrigin: "anonymous"
    }), _react["default"].createElement("script", {
      defer: true,
      dangerouslySetInnerHTML: {
        __html: "\n          (function(sa,fbc){function load(f,c){var a=document.createElement('script');\n          a.async=1;a.src=f;var s=document.getElementsByTagName('script')[0];\n          s.parentNode.insertBefore(a,s);}load(sa);\n          window.addEventListener('load',function(){ \n            window.firebasePerf = firebase.initializeApp(fbc).performance(); \n            __rsfFirebaseSendPerformanceMetrics(window.firebasePerf)\n          });\n          })(".concat(JSON.stringify(firebaseSdkUrl), ", ").concat(JSON.stringify(config), ");")
      }
    }));
  }

  return _react["default"].createElement(_FirebaseContext["default"].Provider, {
    value: context
  }, scripts, children);
}

FirebasePerformanceMonitoring.propTypes = {
  /**
   * Your firebase config.  You can get this from the Firebase console at
   * Settings => General => Your Apps => Firebase SDK Snippet => Config
   */
  config: _propTypes["default"].object.isRequired,

  /**
   * The CDN URL for the standalone firebase SDK
   */
  firebaseSdkUrl: _propTypes["default"].string.isRequired,

  /**
   * The name of the custom trace used to capture client side navigation time.
   * Defaults to "client-side-navigation".
   *
   * To track client side navigation time, you need to add `<TrackPageView/>` to
   * each page component in your app.
   */
  clientSideNavigationTraceName: _propTypes["default"].string
};
FirebasePerformanceMonitoring.defaultProps = {
  clientSideNavigationTraceName: 'client-side-navigation',
  firebaseSdkUrl: 'https://www.gstatic.com/firebasejs/7.15.1/firebase-performance-standalone.js'
};
