// react-storefront-amp/AmpAccordion.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = AmpAccordion;
exports.styles = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _head = _interopRequireDefault(require("next/head"));

var _styles = require("@material-ui/core/styles");

var _amp = require("next/amp");

var _Accordion = _interopRequireDefault(require("react-storefront/Accordion"));

/**
 * @license
 * Copyright © 2017-2018 Moov Corporation.  All rights reserved.
 */
var styles = function styles() {
  return {
    accordion: {}
  };
};

exports.styles = styles;
var useStyles = (0, _styles.makeStyles)(styles, {
  name: 'RSFAmpAccordion'
});

function AmpAccordion(props) {
  var classes = props.classes,
      children = props.children,
      otherProps = (0, _objectWithoutProperties2["default"])(props, ["classes", "children"]);
  classes = useStyles({
    classes: classes
  });

  if ((0, _amp.useAmp)()) {
    return _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement(_head["default"], null, _react["default"].createElement("script", {
      async: true,
      "custom-element": "amp-accordion",
      src: "https://cdn.ampproject.org/v0/amp-accordion-0.1.js"
    })), _react["default"].createElement("amp-accordion", (0, _extends2["default"])({
      "expand-single-section": true,
      "disable-session-states": true,
      "class": classes.accordion
    }, otherProps), _react["default"].Children.map(children, function (child) {
      return _react["default"].cloneElement(child, {
        fromAccordion: true
      });
    })));
  } else {
    return _react["default"].createElement(_Accordion["default"], props);
  }
}
