// react-storefront-amp/AmpTabPanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = AmpTabPanel;
exports.styles = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _TabPanel = _interopRequireDefault(require("react-storefront/TabPanel"));

var _makeStyles = _interopRequireDefault(require("@material-ui/core/styles/makeStyles"));

var _amp = require("next/amp");

var _clsx5 = _interopRequireDefault(require("clsx"));

var _head = _interopRequireDefault(require("next/head"));

var _AmpState = _interopRequireDefault(require("./AmpState"));

var styles = function styles(theme) {
  return {
    ampSelector: {
      '& [option][selected]': {
        outline: 'none',
        cursor: 'inherit'
      }
    },
    ampPanel: {
      display: 'none',
      '&[selected]': {
        display: 'block'
      }
    },
    ampTab: {
      padding: 0,
      opacity: 1,
      '& .label': {
        width: '100%',
        borderBottom: "2px solid transparent",
        opacity: 0.7,
        padding: '13px 12px 11px 12px'
      },
      '& .selected': {
        borderBottom: "2px solid ".concat(theme.palette.secondary.main),
        opacity: 1
      }
    },
    ampTabLabelContainer: {
      padding: 0
    },
    ampTabIndicator: {
      display: 'none'
    }
  };
};

exports.styles = styles;
var useStyles = (0, _makeStyles["default"])(styles, {
  name: 'RSFAmpTabPanel'
});

function AmpTabPanel(_ref) {
  var classes = _ref.classes,
      props = (0, _objectWithoutProperties2["default"])(_ref, ["classes"]);
  classes = useStyles({
    classes: classes
  });
  var amp = (0, _amp.useAmp)();
  return _react["default"].createElement(_react["default"].Fragment, null, amp && _react["default"].createElement(_head["default"], null, _react["default"].createElement("script", {
    async: true,
    "custom-element": "amp-selector",
    src: "https://cdn.ampproject.org/v0/amp-selector-0.1.js"
  })), _react["default"].createElement(_AmpState["default"], {
    initialState: props.selected
  }, function (ampStateId) {
    return _react["default"].createElement(_TabPanel["default"], (0, _extends2["default"])({}, props, {
      tabsProps: {
        classes: {
          indicator: (0, _clsx5["default"])((0, _defineProperty2["default"])({}, classes.ampTabIndicator, amp))
        }
      },
      tabProps: function tabProps(_ref2) {
        var child = _ref2.child,
            index = _ref2.index,
            selected = _ref2.selected;
        return {
          on: "tap:AMP.setState({ ".concat(ampStateId, ": ").concat(index, " })"),
          classes: {
            root: (0, _clsx5["default"])((0, _defineProperty2["default"])({}, classes.ampTab, amp)),
            labelContainer: (0, _clsx5["default"])((0, _defineProperty2["default"])({}, classes.ampTabLabelContainer, amp))
          },
          label: _react["default"].createElement("div", {
            className: (0, _clsx5["default"])('label', {
              selected: selected === index
            }),
            "amp-bind": "class->".concat(ampStateId, " == ").concat(index, " || (").concat(index, " == 0 && !").concat(ampStateId, ") ? 'label selected' : 'label'")
          }, child.props.label)
        };
      },
      panelProps: function panelProps(_ref3) {
        var index = _ref3.index;
        return {
          option: index,
          selected: props.selected === index,
          className: (0, _clsx5["default"])((0, _defineProperty2["default"])({}, classes.ampPanel, amp))
        };
      },
      renderPanels: function renderPanels(panels) {
        return amp ? _react["default"].createElement("amp-selector", {
          role: "tablist",
          "amp-bind": "selected->".concat(ampStateId),
          "class": classes.ampSelector
        }, panels) : panels;
      }
    }));
  }));
}

AmpTabPanel.propTypes = {};
AmpTabPanel.defaultProps = {
  selected: 0
};
