// react-storefront-amp/Video.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = Video;
exports.styles = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _head = _interopRequireDefault(require("next/head"));

var _styles = require("@material-ui/core/styles");

var _amp = require("next/amp");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var styles = function styles(theme) {
  return {
    root: {
      height: '100%',
      width: '100%',
      overflow: 'hidden'
    }
  };
};

exports.styles = styles;
var useStyles = (0, _styles.makeStyles)(styles, {
  name: 'RSFVideo'
});

function Video(_ref) {
  var children = _ref.children,
      classes = _ref.classes,
      props = (0, _objectWithoutProperties2["default"])(_ref, ["children", "classes"]);
  classes = useStyles({
    classes: classes
  });
  var amp = (0, _amp.useAmp)();
  var Tag = amp ? 'amp-video' : 'video';

  var tagProps = _objectSpread((0, _defineProperty2["default"])({}, amp ? 'class' : 'className', classes.root), props); // Layout must be declared in AMP, so we use "fill" by default if not defined


  if (amp && !props.height && !props.width && !props.layout) {
    tagProps.layout = 'fill';
  }

  return _react["default"].createElement(_react["default"].Fragment, null, amp && _react["default"].createElement(_head["default"], null, _react["default"].createElement("script", {
    async: true,
    "custom-element": "amp-video",
    src: "https://cdn.ampproject.org/v0/amp-video-0.1.js"
  })), _react["default"].createElement(Tag, tagProps, children));
}

Video.propTypes = {
  /**
   * Set to false to hide the browser video controls
   */
  controls: _propTypes["default"].bool
};
Video.defaultProps = {
  controls: true
};
